netx/dialer/quicsaver.go

```diff
@@ -0,0 +1,126 @@
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"time"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/ooni/probe-engine/internal/tlsx"
+	"github.com/ooni/probe-engine/netx/errorx"
+	"github.com/ooni/probe-engine/netx/trace"
+)
+
+// QUICSaverDialer saves events occurring during the dial
+type QUICSaverDialer struct {
+	QUICContextDialer
+	Saver *trace.Saver
+}
+
+// DialContext implements Dialer.DialContext
+func (d QUICSaverDialer) DialContext(ctx context.Context, network, addr string, host string, tlsCfg *tls.Config, cfg *quic.Config) (quic.EarlySession, error) {
+	start := time.Now()
+	sess, err := d.QUICContextDialer.DialContext(ctx, network, addr, host, tlsCfg, cfg)
+	stop := time.Now()
+	d.Saver.Write(trace.Event{
+		Address:  host,
+		Duration: stop.Sub(start),
+		Err:      err,
+		Name:     errorx.QUICHandshakeOperation,
+		Proto:    network,
+		Time:     stop,
+	})
+	return sess, err
+}
+
+// QUICHandshakeSaver saves events occurring during the handshake
+type QUICHandshakeSaver struct {
+	Saver  *trace.Saver
+	Dialer QUICContextDialer
+}
+
+// DialContext implements QUICContextDialer.DialContext
+func (h QUICHandshakeSaver) DialContext(ctx context.Context, network string, addr string, host string, tlsCfg *tls.Config, cfg *quic.Config) (quic.EarlySession, error) {
+	start := time.Now()
+	h.Saver.Write(trace.Event{
+		Name:          "quic_handshake_start",
+		NoTLSVerify:   tlsCfg.InsecureSkipVerify,
+		TLSNextProtos: tlsCfg.NextProtos,
+		TLSServerName: tlsCfg.ServerName,
+		Time:          start,
+	})
+	sess, err := h.Dialer.DialContext(ctx, network, addr, host, tlsCfg, cfg)
+	stop := time.Now()
+
+	if sess == nil {
+		h.Saver.Write(trace.Event{
+			Duration:      stop.Sub(start),
+			Err:           err,
+			Name:          "quic_handshake_done",
+			NoTLSVerify:   tlsCfg.InsecureSkipVerify,
+			TLSNextProtos: tlsCfg.NextProtos,
+			TLSServerName: tlsCfg.ServerName,
+			Time:          stop,
+		})
+		return sess, err
+	}
+	state := tls.ConnectionState{}
+	if sess != nil {
+		state = ConnectionState(sess)
+	}
```

Comment:
I think we've already excluded the sess == nil case?

Reply:
right, but not with the comment above anymore :) unless sess == nil and err != nil are completely synonym here, which I'm not too confident about

Reply:
This is where a language with a sum type would have helped. The semantics of this code is that we either have an error or the other thing is valid. It's used quite extensively in Go and also in probe-engine.

I agree that it would be much nicer to have `ErrorOr<Seession>` or `Error|Session` (even better), but sadly we're using a less advanced language where we humans need to do the sum type manually :^).

Reply:
Still, if you want to enforce the sum type semantics here, you should intercept the case where error is nil and also session is nil and return yourself an error (there is code in the Go HTTP standard library that does that).

I think we should not write code this defensive though.